src/cunumeric/stat/histogram.cuh

```diff
@@ -0,0 +1,148 @@
+/* Copyright 2023 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+
+#include <thrust/binary_search.h>
+#include <thrust/device_vector.h>
+#include <thrust/fill.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/sort.h>
+
+#include <thrust/reduce.h>
+#include <thrust/scatter.h>
+
+#include <thrust/execution_policy.h>
+#include <thrust/system/omp/execution_policy.h>
+#include <cuda_runtime.h>
+#include <thrust/version.h>
+
+#include <cub/cub.cuh>
+
+#ifdef _DEBUG
+#include <cassert>
+#include <iostream>
+#endif
+
+#include "cunumeric/utilities/thrust_util.h"
+
+#include "cunumeric/stat/histogram_gen.h"
+
+namespace cunumeric {
+namespace detail {
+template <typename element_t>
+decltype(auto) get_raw_ptr(Buffer<element_t>& v)
+{
+  return v.ptr(0);
+}
+
+// rudimentary device allocator:
+//
+template <typename elem_t, typename exe_policy_t>
+struct allocator_t<
+  elem_t,
+  exe_policy_t,
+  std::enable_if_t<!std::is_same_v<exe_policy_t, thrust::detail::host_t> &&
+                   !std::is_same_v<exe_policy_t, thrust::system::omp::detail::par_t>>> {
+  allocator_t(void) {}
+
+  allocator_t(elem_t, exe_policy_t) {}  // tag-dispatch for CTAD
+
+  elem_t* operator()(exe_policy_t exe_pol, size_t size, elem_t init = 0)
+  {
+    d_buffer_     = create_buffer<elem_t>(size, Legion::Memory::Kind::GPU_FB_MEM);
+    elem_t* d_ptr = get_raw_ptr(d_buffer_);
+
+    thrust::fill_n(exe_pol, d_ptr, size, init);
+
+    return d_ptr;
+  }
+
+ private:
+  Buffer<elem_t> d_buffer_;
+};
+
+template <typename exe_policy_t>
+void synchronize_exec(exe_policy_t, cudaStream_t stream)
+{
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+}
+
+// device version:
+//
+template <typename exe_policy_t, typename weight_t, typename offset_t, typename allocator_t>
+struct segmented_sum_t<
+  exe_policy_t,
+  weight_t,
+  offset_t,
+  allocator_t,
+  std::enable_if_t<!std::is_same_v<exe_policy_t, thrust::detail::host_t> &&
+                   !std::is_same_v<exe_policy_t, thrust::system::omp::detail::par_t>>> {
```

Comment:
I have to pass those different execution  policies to the thrust algorithms. So, if I use `VariantKind` I then need to create an additional compile time mapping between `VariantKind` -> `exec_policy_t`. Why not specializing directly on `exec_policy_t`, then? Since I need it / sue it, while `VariantKind` is not really needed / used it here.

Reply:
OK, in that case let's not use the internal thrust types. The thrust documentation [states](https://thrust.github.io/doc/group__execution__policies_ga330d16be82b0518b6a37d026c2811e7e.html#ga330d16be82b0518b6a37d026c2811e7e) that:

> The type of [thrust::device](https://thrust.github.io/doc/group__execution__policies_ga330d16be82b0518b6a37d026c2811e7e.html#ga330d16be82b0518b6a37d026c2811e7e) is implementation-defined.